---
params:
  title: "Genomic Array Quality-Control"
  author_name: "Firstname Lastname"
  author_affiliation: "Institution"
  author_email: "some@email.com"
  input_files: NULL
  output_directory: NULL
  cohort_name: NULL
  array: NULL
  callrate_samples: 0.95
  callrate_snps: 0.95
  sex_threshold: !r c(0.2, 0.8)
  heterozygosity_threshold: 4
  maf_threshold: 0.01
  hwe_pvalue: 0.0001
  includes_relatives: FALSE
  mendelian_samples: 0.05
  mendelian_snps: 0.1
  ibd_threshold: 0.2
  population: NULL
  pca_components: 10
  pca_threshold: 3
  max_labels: 15
  check_bim_script: !r system.file("perl", "HRC-1000G-check-bim-NoReadKey.pl", package = "dgapaq")
  imputation_ref: "1KG"
  imputation_panel: "/disks/DATA/ExternalData/1kg/hg19/Reference_genome/1000GP_Phase3_combined.legend"
  ref1kg_panel: "/disks/DATA/ExternalData/1kg/samples_description/integrated_call_samples_v3.20130502.ALL.panel"
  ref1kg_fasta: "/disks/DATA/ExternalData/1kg/hg19/human_g1k_v37.fasta"
  ref1kg_population: "/disks/DATA/ExternalData/1kg/samples_description/1kg_pop_description.tsv"
  ref1kg_genotypes: "/disks/DATA/ExternalData/1kg/hg19/Genotypes/RawPLINK/ALL/"
  bin_path: !r list(bcftools = "/usr/bin/bcftools", bgzip = "/usr/bin/bgzip", plink = "/usr/bin/plink1.9", gcta = "/usr/bin/gcta64")
title: '`r params[["title"]]`'
author:
- name: '`r params[["author_name"]]`'
  affiliation: '`r params[["author_affiliation"]]`'
  email: '`r params[["author_email"]]`'
date: '`r format(Sys.time(), "%B %d, %Y")`'
output:
  bookdown::html_document2:
    theme: simplex
    toc: true
    toc_depth: 3
    toc_float: 
      collapsed: false
    fig_width: 6.3
    fig_height: 4.7
    number_sections: true
    self_contained: true
editor_options: 
  chunk_output_type: console
---

```{r setup, include = FALSE}
options(stringsAsFactors = FALSE)


### Environment ====================================================================================
output_directory <- params[["output_directory"]]
invisible(
  sapply(
    X = file.path(output_directory, c("tmp", "plink_qc", "vcf_qc")), 
    FUN = dir.create,
    showWarnings = FALSE, 
    recursive = TRUE, 
    mode = "0777"
  )
)
temp <- file.path(output_directory, "tmp")
dir.create(file.path(temp, "1kgenome_pruned"), showWarnings = FALSE, recursive = TRUE, mode = "0777")

plink <- params[["bin_path"]][["plink"]]


### Load packages ==================================================================================
suppressPackageStartupMessages({
  library("bookdown")
  library("data.table")
  library("ggplot2")
  library("ggforce")
  library("ggrepel")
  library("ggtext")
  library("gt")
  library("knitr")
  library("parallel")
  library("patchwork")
  library("scales")
  library("stats")
  library("sessioninfo")
  library("writexl")
})


### knitr settings =================================================================================
knitr::opts_chunk$set(
  results = "asis",
  size = "small",
  include = TRUE,
  echo = FALSE,
  warning = FALSE,
  message = FALSE,
  dpi = 120,
  tidy = FALSE,
  crop = TRUE,
  autodep = TRUE,
  fig.align = "center",
  fig.pos = "!H"
)


### Define theme ===================================================================================
ggplot2::theme_set(
  ggplot2::theme_light(base_size = 11) +
    ggplot2::theme(
      plot.title = ggtext::element_markdown(), 
      plot.title.position = "plot", 
      plot.subtitle = ggtext::element_markdown(face = "italic", size = ggplot2::rel(0.8)),
      plot.caption = ggtext::element_markdown(face = "italic", size = ggplot2::rel(0.5)), 
      plot.caption.position = "plot",
      axis.text.x = ggtext::element_markdown(), 
      axis.text.x.top = ggtext::element_markdown(), 
      axis.text.y = ggtext::element_markdown(), 
      legend.box.just = "left", 
      legend.key.size = ggplot2::unit(0.5, "lines"), 
      legend.text = ggtext::element_markdown(size = ggplot2::rel(0.5))
    )
)


### Functions ======================================================================================
sys <- function(command) system(command = command, ignore.stdout = TRUE, ignore.stderr = TRUE)

`%>%` <- gt::`%>%`

percent1 <- scales::percent_format(accuracy = 0.1, suffix = " %")
percent2 <- scales::percent_format(accuracy = 0.01, suffix = " %")
comma <- scales::comma_format(accuracy = 1)
scientific <- function(x, ...) gsub("(.*)e([-+]*)0*(.*)", "\\1 &times; 10<sup>\\2\\3</sup>", scales::scientific(x, ...))

pval_trans <- function(alpha = NULL, md = FALSE, prefix = FALSE) {
  scales::trans_new(
    name = "pval",
    domain = c(0, 1),
    transform = function(x) {x[x < .Machine$double.xmin] <- .Machine$double.xmin; -log(x, 10)},
    inverse = function(x) {10^-x},
    breaks = (function(n = 5) {
      function(x) {
        max <- floor(-log(min(c(x, alpha), na.rm = TRUE), base = 10))
        if (max == 0) 1 else sort(unique(c(10^-seq(0, max, by = floor(max / n) + 1), alpha)))
      }
    })(),
    format = (function(x) {
      g <- function(x) {
        gsub(
          "1 %*% ", "",
          gsub(
            "(.*)e([-+]*)0*(.*)", "\\1 %*% 10^\\2\\3",
            gsub(
             "1e+00", "1",
              scales::scientific(x),
             fixed = TRUE
            )),
          fixed = TRUE
        )
      }
      highlight_alpha <- function(x, md = FALSE, prefix = FALSE) {
        if (md & nchar(system.file(package = "ggtext")) != 0) {
          prefix_text <- if (prefix) "&alpha; = " else ""
          out <- paste0(
            "<b style = 'color:firebrick2;'>", 
            gsub("(.*) [%][*][%] .*\\^(.*)", paste(prefix_text, "\\1 &times; 10<sup>\\2</sup>"), g(x)), 
            "</b>"
          )
        } else {
          prefix_text <- if (prefix) "alpha ==" else ""
          out <- parse(text = paste(prefix_text, g(x)))
        }
        
        out
      }
      
      if (!is.null(alpha)) {
        ifelse(
          test = scales::scientific(x) == scales::scientific(alpha), 
          yes = highlight_alpha(x, md, prefix), 
          no = g(x)
        )
      } else {
        parse(text = g(x))
      }
    })
  )
}
```

```{r data}
if (length(list.files(dirname(params[["input_files"]]), pattern = ".bed")) != 0) {
  sys(paste(plink, 
    "--bfile", params[["input_files"]], 
    "--snps-only", 
    "--make-bed",
    "--out", file.path(temp, "data")
  ))
} else {
  sys(paste(plink, 
    "--file", params[["input_files"]], 
    "--snps-only", 
    "--make-bed",
    "--out", file.path(temp, "data")
  ))
}
```

# Parameters

* Raw data: 
    `` `r params[["input_files"]]` ``
* 1,000 Genomes files: 
    `` `r params[["ref1kg_genotypes"]]` ``
* Reference population: 
    `` `r if (is.null(params[["population"]])) "unspecified" else params[["population"]]` ``
* Sample-based call rate threshold: 
    `` `r params[["callrate_samples"]]` ``
* Variant-based call rate threshold: 
    `` `r params[["callrate_snps"]]` ``
* Number of standard deviations (SD) from the mean heterozygosity rate: 
    `` `r params[["heterozygosity_threshold"]]` ``
* Proportion of identity by descent (IBD) threshold: 
    `` `r params[["ibd_threshold"]]` ``
* Number of components from the principal component analysis (PCA): 
    `` `r params[["pca_components"]]` ``
* Number of standard deviations (SD) from the population centroid based on the PCA: 
    `` `r params[["pca_threshold"]]` ``
* Minor allele frequency (MAF) threshold: 
    `` `r params[["maf_threshold"]]` ``
* Hardy-Weinberg equilibrium p-value: 
    `` `r scales::scientific(params[["hwe_pvalue"]])` ``
* Mendelian error rate threshold:
    + Maximum per-trio error rate: 
        `` `r if (params[["includes_relatives"]]) params[["mendelian_samples"]] else "not applicable"` `` 
    + Ma per-variant error rate: 
        `` `r if (params[["includes_relatives"]]) params[["mendelian_snps"]] else "not applicable"` `` 

# Cohort Overview

Genotyped using `r if (is.null(params[["array"]])) "unspecified" else params[["array"]]` array.

```{r cohort-overview, include = FALSE}
fam_data <- data.table::fread(
  file = file.path(temp, "data.fam"), 
  col.names = c("FID", "IID", "father_id", "mother_id", "sex", "phenotype"),
  colClasses = c("character", "character", "character", "character", "numeric", "numeric")
)

samples_duplicated <- sum(table(gsub("[_-][^-_]*$", "", fam_data[["IID"]])) > 1)

snps_number <- nrow(data.table::fread(input = file.path(temp, "data.bim"), select = 1))

fam_data[, cohort := params[["cohort_name"]]]
fam_data[, sex_fct := factor(sex, levels = c(1, 2, 0), labels = c("Male", "Female", "Unspecified"))]
```

PLINK files includes:

* **`r comma(nrow(fam_data))`** samples.  
    + **Note:** **`r comma(samples_duplicated)`** samples seemed to have duplicates.
* **`r comma(snps_number)`** variants.

```{r cohort-overview-table}
fam_data[
  j = .N, 
  by = .(FID, gsub("[_-][^-_]*$", "", IID), cohort, sex_fct)
][
  j = .N, by = c("cohort", "sex_fct")
][
  j = .(sex_fct, N)
] %>% 
  gt::gt(auto_align = "center") %>%
  gt::tab_header(
    title = "Samples Available", 
    subtitle = gt::md(paste0("*", params[["cohort_name"]], "*"))
  ) %>%
  gt::fmt_number(columns = "N", decimals = 0) %>% 
  gt::grand_summary_rows(
    columns = gt::vars(N),
    fns = list(Total = ~sum(.)),
    formatter = gt::fmt_number,
    decimals = 0
  ) %>% 
  gt::cols_label(sex_fct = "Sex") %>% 
  gt::opt_row_striping() %>% 
  gt::opt_all_caps() %>% 
  print()
```

# Sample-based QC {#sample-qc}

## Sample Call Rate

Call rate was computed using the `--missing` command in PLINK. 

```{r sample-call-rate}
sys(paste(plink, 
  "--bfile", file.path(temp, "data"), 
  "--missing", 
  "--out", file.path(temp, "data")
))

callrate_samples <- data.table::fread(
  file = file.path(temp, "data.imiss"), 
  colClasses = c("FID" = "character", "IID" = "character")
)[, labs := data.table::fifelse(F_MISS > 1 - params[["callrate_samples"]], IID, NA_character_)][order(F_MISS)]
samples_to_exclude <- callrate_samples[!is.na(labs), .(FID, IID)]
samples_to_exclude[, QC := "Sample_Call_Rate"]

sys(paste(plink, 
  "--bfile", file.path(temp, "data"), 
  "--freq", 
  "--out", file.path(temp, "data")
))

freqsnp <- data.table::fread(file.path(temp, "data.frq"))

data.table::fwrite(
  x = freqsnp[MAF >= params[["maf_threshold"]], .(SNP)],
  file = file.path(temp, "data_maf_geq.txt"),
  col.names = FALSE, 
  row.names = FALSE, 
  quote = FALSE
)

data_maf_lower <- freqsnp[MAF < params[["maf_threshold"]], .(SNP)]
is_maf_lower <- nrow(data_maf_lower) > 0

sys(paste(plink, 
  "--bfile", file.path(temp, "data"), 
  "--missing", 
  "--extract", file.path(temp, "data_maf_geq.txt"), 
  "--out", file.path(temp, "data_maf_geq")
))

callrate_samples_geq <- data.table::fread(
  file = file.path(temp, "data_maf_geq.imiss"),
  colClasses = c("FID" = "character", "IID" = "character")
)[order(F_MISS)]
prop_all <- sum(callrate_samples[, "F_MISS"] > 1 - params[["callrate_samples"]]) / nrow(callrate_samples)
prop_maf_geq <- sum(callrate_samples_geq[, "F_MISS"] > 1 - params[["callrate_samples"]]) / nrow(callrate_samples_geq)
prop_diff_maf_geq <- abs((prop_maf_geq - prop_all) / prop_all) >= 0.05

if (is_maf_lower) {
  data.table::fwrite(
    x = data_maf_lower,
    file = file.path(temp, "data_maf_lower.txt"),
    col.names = FALSE, 
    row.names = FALSE, 
    quote = FALSE
  )
  sys(paste(plink, 
    "--bfile", file.path(temp, "data"), 
    "--missing", 
    "--extract", file.path(temp, "data_maf_lower.txt"), 
    "--out", file.path(temp, "data_maf_lower")
  ))
  callrate_samples_lower <- data.table::fread(
    file = file.path(temp, "data_maf_lower.imiss"),
    colClasses = c("FID" = "character", "IID" = "character")
  )[order(F_MISS)]
  prop_maf_lower <- sum(callrate_samples_lower[, "F_MISS"] > 1 - params[["callrate_samples"]]) / nrow(callrate_samples_lower)
  prop_diff_maf_lower <- abs((prop_maf_lower - prop_all) / prop_all) >= 0.05
} else {
  prop_diff_maf_lower <- NA
}

callrate_samples_exclude <- callrate_samples[F_MISS > 1 - params[["callrate_samples"]]][order(F_MISS)]
callrate_samples_exclude[, Call_Rate := 1 - F_MISS]
callrate_samples_exclude <- callrate_samples_exclude[, .(FID, IID, Call_Rate)]

cat(
  "\n\nA call rate lower than ", percent2(params[["callrate_samples"]]), " was observed for **",
  comma(nrow(callrate_samples_exclude)), "** samples, leading to their exclusion.\n\n",
  sep = ""
)
```

```{r sample-call-rate-tab}
cols_table <- c(
  "Call Rate Threshold", 
  "Samples to Exclude\n(All Variants)", 
  paste0("Samples to Exclude\n(Variants with MAF ≥ ", params[["maf_threshold"]], ")")
)

callrate_thresholds <- sort(unique(c(params[["callrate_samples"]], 0.90, 0.95, 0.97, 0.98, 0.99)), decreasing = TRUE)

cbind.data.frame(
  percent2(callrate_thresholds),
  rowSums(sapply(callrate_samples[["F_MISS"]], ">", 1 - callrate_thresholds)),
  rowSums(sapply(callrate_samples_geq[["F_MISS"]], ">", 1 - callrate_thresholds))
) %>% 
  `colnames<-`(cols_table) %>% 
  gt::gt(auto_align = "center") %>% 
  gt::tab_header("Number Of Samples To Exclude Based On Call Rate Thresholds") %>% 
  gt::tab_style(
    style = gt::cell_fill(color = "dodgerblue", alpha = 0.5), 
    locations = gt::cells_body(
      columns = gt::vars(!!!cols_table), 
      rows = `Call Rate Threshold` == percent2(params[["callrate_samples"]])
    )
  ) %>% 
  gt::opt_row_striping() %>% 
  gt::opt_all_caps() %>% 
  print()

cat("\n")
```

```{r sample-call-rate-fig}
ggplot2::ggplot(
  data = callrate_samples,
  mapping = ggplot2::aes(x = length(F_MISS):1, y = 1 - F_MISS)
) +
  ggplot2::geom_point(
    colour = scales::viridis_pal(begin = 0.5, end = 0.5)(1), 
    shape = 1, 
    na.rm = TRUE
  ) +
  ggrepel::geom_label_repel(
    data = ~ .x[, labs := if (sum(!is.na(labs)) > params[["max_labels"]]) NA else labs],
    mapping = ggplot2::aes(label = labs),
    min.segment.length = ggplot2::unit(0, "lines"),
    force = 10, 
    fill = "white",
    colour  = "firebrick2", 
    segment.colour = "firebrick2", 
    size = 3, 
    na.rm = TRUE
  ) +
  ggplot2::geom_hline(
    mapping = ggplot2::aes(yintercept = params[["callrate_samples"]]),
    colour = "firebrick2", 
    linetype = 2
  ) +
  ggplot2::scale_x_continuous(labels = comma, trans = "log10") +
  ggplot2::scale_y_continuous(
    breaks = function(x) unique(c(scales::breaks_extended()(x), params[["callrate_samples"]])),
    labels = function(x) ifelse(x == params[["callrate_samples"]], paste0("<b style='color:firebrick2;'>", percent1(x), "</b>"), percent1(x))
  ) +
  ggplot2::labs(x = "Number of Samples", y = "Call Rate", title = "Sample Call Rate")
```

```{r sample-call-rate-exclude, eval = nrow(callrate_samples_exclude) != 0}
callrate_samples_exclude %>% 
  gt::gt(auto_align = "center") %>% 
  gt::fmt_percent("Call_Rate", incl_space = TRUE) %>% 
  gt::cols_label(Call_Rate = "Call Rate") %>% 
  gt::tab_header(paste(
    "Samples To Be Excluded For A Call Rate Threshold Of", percent2(params[["callrate_samples"]])
  )) %>% 
  gt::opt_row_striping() %>% 
  gt::opt_all_caps() %>% 
  print()
```

```{r sample-call-rate-note}
is_cr_note <- (!is.na(prop_diff_maf_lower) & prop_diff_maf_lower) | 
  (!is.na(prop_diff_maf_geq) & prop_diff_maf_geq)
if (is_cr_note) {
  cat("**Note:**\n\n")
  if (!is.na(prop_diff_maf_lower) & prop_diff_maf_lower) {
    cat(
      "* The sample-based call rate was different for ", 
      "variants with MAF < ", params[["maf_threshold"]],
      " (", percent2(prop_maf_lower), ") ", 
      "compared to the call rate based on all variants ", 
      "(", percent2(prop_all), ").\n\n", 
      sep = ""
    )
  }
  
  if (!is.na(prop_diff_maf_geq) & prop_diff_maf_geq) {
    cat(
      "* The sample-based call rate was different for ",
      "variants with MAF ≥ ", params[["maf_threshold"]], 
      " (", percent2(prop_maf_geq), ") ", 
      "compared to the call rate based on all variants ",
      "(", percent2(prop_all), ").\n\n", 
      sep = ""
    )
  }
  cat("\n")
}
```

## Genotypic Sex (Homozygosity Rate)

The mean homozygosity rate across the X-chromosome was computed using the `--check-sex` command in PLINK.

```{r sex-check}
sys(paste(plink, 
  "--bfile", file.path(temp, "data"), 
  "--check-sex", min(params[["sex_threshold"]]), max(params[["sex_threshold"]]),
  "--out", file.path(temp, "data")
))

sex_file_check <- file.exists(file.path(temp, "data.sexcheck"))

if (!sex_file_check) {
  cat("**Warning:** sex check was not achieved, *i.e.*, no polymorphic X chromosome.\n\n")
}

sexcheck <- merge(
  x = data.table::fread(
    file = file.path(temp, "data.sexcheck"), 
    colClasses = c("FID" = "character", "IID" = "character")
  ),
  y = callrate_samples, 
  by = c("FID", "IID"), 
  all = TRUE
)
sexcheck <- merge(x = sexcheck, y = fam_data, by = c("FID", "IID"), all.x = TRUE)
sexcheck[, sex_discrepancy := sex != SNPSEX]
sexcheck[, STATUS := data.table::fifelse(sex_discrepancy | is.na(sex_discrepancy) | `F` >= min(params[["sex_threshold"]]) & `F` < max(params[["sex_threshold"]]), "PROBLEM", STATUS)]
sexcheck[, labs := data.table::fifelse(STATUS == "PROBLEM" & F_MISS < 1 - params[["callrate_samples"]], IID, NA_character_)]

sexcheck_problem <- sexcheck[STATUS == "PROBLEM"]

sexcheck_discrepancy <- sexcheck_problem[(sex_discrepancy), .(FID, IID)][, QC := "Sex_Discrepancy"]
sexcheck_missing <- sexcheck_problem[is.na(sex_discrepancy), .(FID, IID)][, QC := "Sex_Missing"]
sexcheck_other <- sexcheck_problem[!(sex_discrepancy), .(FID, IID)][, QC := "Sex_F"] 

samples_to_exclude <- rbind(
  samples_to_exclude,
  sexcheck_discrepancy,
  sexcheck_missing,
  sexcheck_other
)

if (sex_file_check & nrow(sexcheck_problem) > 0) {
  cat(
    paste0(
      'For **', comma(nrow(sexcheck_problem)), '** samples ', 
      '(including **', comma(sum(sexcheck_problem[, "sex"] == 0)), '**, with unspecified sex and **', comma(sum(sexcheck_problem[, "SNPSEX"] == 0)), '** samples with undetermined genotypic sex)',
      ', the sex discrepancy between genotyping data and reported sex could not be resolved.',
      '\n\n'
    )
  )
}
```

```{r sex-check-fig, eval = sex_file_check}
ggplot2::ggplot(
  data = sexcheck[!is.na(`F`), ],
  mapping = ggplot2::aes(x = `F`, y = 1 - F_MISS)
) +
  ggplot2::geom_rect(
    mapping = ggplot2::aes(xmin = -Inf, xmax = Inf, ymin = -Inf, ymax = params[["callrate_samples"]]), 
    fill = "grey95",
    colour = "transparent"
  ) +
  ggplot2::annotate(
    geom = "rect",
    xmin = -Inf, xmax = -0.2, ymin = -Inf, ymax = Inf, 
    fill = "firebrick2",
    alpha = 0.10,
    colour = "transparent"
  ) +
  ggplot2::annotate(
    geom = "rect",
    xmin = -0.2, xmax = min(params[["sex_threshold"]]), ymin = -Inf, ymax = Inf, 
    fill = "firebrick2",
    alpha = 0.2,
    colour = "transparent"
  ) +
  ggplot2::annotate(
    geom = "rect",
    xmin = max(params[["sex_threshold"]]), xmax = 1, ymin = -Inf, ymax = Inf, 
    fill = "dodgerblue",
    alpha = 0.2,
    colour = "transparent"
  ) +
  ggplot2::geom_hline(yintercept = params[["callrate_samples"]], colour = "firebrick2", linetype = 2) +
  ggplot2::geom_point(
    data = ~ .x[!(sex_discrepancy)], 
    na.rm = TRUE, 
    shape = 1,
    colour = scales::viridis_pal(begin = 0.5, end = 0.5)(1)
  ) +
  ggplot2::geom_point(
    data = ~ .x[(sex_discrepancy)], 
    na.rm = TRUE, 
    shape = 3,
    colour  = "firebrick2"
  ) +
  ggrepel::geom_label_repel(
    data = ~ .x[(sex_discrepancy)][, labs := if (sum(sex_discrepancy) > params[["max_labels"]]) NA else labs],
    mapping = ggplot2::aes(label = labs),
    min.segment.length = ggplot2::unit(0, "lines"),
    force = 10, 
    fill = "white",
    colour  = "firebrick2", 
    segment.colour = "firebrick2", 
    size = 3, 
    na.rm = TRUE
  ) +
  ggplot2::scale_x_continuous(breaks = c(0, params[["sex_threshold"]], 1), labels = percent1) +
  ggplot2::scale_y_continuous(
    breaks = function(x) unique(c(scales::breaks_extended()(x), params[["callrate_samples"]])),
    labels = function(x) ifelse(x == params[["callrate_samples"]], paste0("<b style='color:firebrick2;'>", percent1(x), "</b>"), percent1(x))
  ) +
  ggplot2::labs(
    x = "Homozygosity Rate",
    y = "Call Rate",
    title = "Homozygosity Rate Based on X-chromosome Variants",
    subtitle = paste(
      "The expected homozygosity rate range:",
      "<b style = 'color:firebrick2;'>for female</b> and", 
      "<b style = 'color:dodgerblue;'>for male</b>.<br>", 
      "Samples for which sex information is discordant or missing are denoted by \"<b style = 'color:firebrick2;'>+</b>\"."
    )
  ) + 
  ggplot2::theme(legend.position = "none")
```

```{r sex-check-tab, eval = sex_file_check}
if (nrow(sexcheck_problem) > 0) {
  sexcheck_problem[(sex_discrepancy), .(FID, IID, Phenotype = PEDSEX, Genotype = SNPSEX)] %>% 
    gt::gt(auto_align = "center") %>% 
    gt::tab_header("Samples To Be Checked") %>% 
    gt::opt_row_striping() %>% 
    gt::opt_all_caps() %>% 
    print()
}
```

## Heterozygosity

The observed heterozygosity rate per sample was computed using the formula:  
(Number of non-missing genotypes N(NM) - Number of homozygous genotypes O(Hom)) / N(NM).

N(NM) and O(Hom) were obtained using the `--het` command in PLINK.

```{r heterozygosity, include = FALSE}
sys(paste(plink, 
  "--bfile", file.path(temp, "data"), 
  "--het",
  "--out", file.path(temp, "data_het")
))

heterozygosity <- merge(
  x = data.table::fread(
    file = file.path(temp, "data_het.het"),
    colClasses = c("FID" = "character", "IID" = "character"), 
    check.names = TRUE
  ), 
  y = callrate_samples, 
  by = c("FID", "IID"), 
  all = TRUE
)
heterozygosity[, hrate := (N.NM. - O.HOM.) / N.NM., by = F_MISS < 1 - params[["callrate_samples"]]]
heterozygosity[, colour := sqrt( (hrate - mean(hrate, na.rm = TRUE))^2 + (F_MISS - mean(F_MISS, na.rm = TRUE))^2 ), by = F_MISS < 1 - params[["callrate_samples"]]]
heterozygosity[, 
  labs := data.table::fifelse(
    test = F_MISS < 1 - params[["callrate_samples"]] & ( 
      hrate > mean(hrate, na.rm = TRUE) + params[["heterozygosity_threshold"]] * stats::sd(hrate, na.rm = TRUE) | 
        hrate < mean(hrate, na.rm = TRUE) - params[["heterozygosity_threshold"]] * stats::sd(hrate, na.rm = TRUE)
    ), 
    yes = IID, 
    no = NA_character_
  ), 
  by = F_MISS < 1 - params[["callrate_samples"]]
]

samples_to_exclude <- rbind(
  samples_to_exclude,
  heterozygosity[!is.na(labs), .(FID, IID)][, QC := "Heterozygosity_Check"]
)

if (is_maf_lower) {
  sys(paste(plink, 
    "--bfile", file.path(temp, "data"),
    "--extract", file.path(temp, "data_maf_lower.txt"),
    "--het",
    "--out", file.path(temp, "data_het_maf_lower")
  ))

  heterozygosity_maf_lower <- merge(
    x = data.table::fread(
      file = file.path(temp, "data_het_maf_lower.het"),
      colClasses = c("FID" = "character", "IID" = "character"), 
      check.names = TRUE
    ), 
    y = callrate_samples, 
    by = c("FID", "IID"), 
    all = TRUE
  )
  heterozygosity_maf_lower[, hrate := (N.NM. - O.HOM.) / N.NM., by = F_MISS < 1 - params[["callrate_samples"]]]
  heterozygosity_maf_lower[, 
    colour := sqrt( (hrate - mean(hrate, na.rm = TRUE))^2 + (F_MISS - mean(F_MISS, na.rm = TRUE))^2 ), 
    by = F_MISS < 1 - params[["callrate_samples"]]
  ]
  heterozygosity_maf_lower[, 
    labs := data.table::fifelse(
      test = F_MISS < 1 - params[["callrate_samples"]] & ( 
        hrate > mean(hrate, na.rm = TRUE) + params[["heterozygosity_threshold"]] * stats::sd(hrate, na.rm = TRUE) | 
          hrate < mean(hrate, na.rm = TRUE) - params[["heterozygosity_threshold"]] * stats::sd(hrate, na.rm = TRUE)
      ), 
      yes = IID, 
      no = NA_character_
    ), 
    by = F_MISS < 1 - params[["callrate_samples"]]
  ]

  samples_to_exclude <- rbind(
    samples_to_exclude,
    heterozygosity_maf_lower[!is.na(labs), .(FID, IID)][, QC := paste0("Heterozygosity_Check_maf_", params[["maf_threshold"]])]
  )
}
```

```{r heterozygosity-fig}
plot_het <- function(data, params) {
  ggplot2::ggplot(data = data[, labs := if (sum(!is.na(labs)) > params[["max_labels"]]) NA else labs], mapping = ggplot2::aes(x = 1 - F_MISS, y = hrate)) +
    ggplot2::geom_rect(
      mapping = ggplot2::aes(ymin = -Inf, ymax = Inf, xmin = -Inf, xmax = params[["callrate_samples"]]), 
      fill = "grey95",
      colour = "transparent"
    ) +
    ggplot2::geom_point(mapping = ggplot2::aes(colour = colour), shape = 1, na.rm = TRUE) +
    ggplot2::geom_vline(
      mapping = ggplot2::aes(xintercept = params[["callrate_samples"]]),
      colour = "firebrick2", 
      linetype = 2
    ) +
    ggplot2::geom_hline(
      data = ~ .x[
        F_MISS < 1 - params[["callrate_samples"]], 
        .(
          yintercept = mean(hrate, na.rm = TRUE) - params[["heterozygosity_threshold"]] * stats::sd(hrate, na.rm = TRUE)
        )
      ],
      mapping = ggplot2::aes(yintercept = yintercept),
      colour = "firebrick2", 
      linetype = 2
    ) +
    ggplot2::geom_hline(
      data = ~ .x[
        F_MISS < 1 - params[["callrate_samples"]], 
        .(
          yintercept = mean(hrate, na.rm = TRUE) + params[["heterozygosity_threshold"]] * stats::sd(hrate, na.rm = TRUE)
        )
      ],
      mapping = ggplot2::aes(yintercept = yintercept),
      colour = "firebrick2", 
      linetype = 2
    ) +
    ggrepel::geom_label_repel(
      data = ~ .x[!is.na(labs)],
      mapping = ggplot2::aes(label = labs),
      min.segment.length = ggplot2::unit(0, "lines"),
      force = 10, 
      fill = "white",
      colour  = "firebrick2", 
      segment.colour = "firebrick2", 
      size = 3, 
      na.rm = TRUE
    ) + 
    ggplot2::scale_colour_viridis_c(begin = 0.2, end = 0.8, trans = "sqrt") +
    ggplot2::scale_x_continuous(
      breaks = function(x) unique(c(scales::breaks_extended()(x), params[["callrate_samples"]])),
      labels = function(x) ifelse(x == params[["callrate_samples"]], paste0("<b style='color:firebrick2;'>", percent1(x), "</b>"), percent1(x)),
      guide = ggplot2::guide_axis(angle = 45)
    ) +
    ggplot2::labs(x = "Call Rate", y = "Heterozygosity Rate") +
    ggplot2::theme(legend.position = "none")
}

if (is_maf_lower) {
  patchwork::wrap_plots(plot_het(heterozygosity, params), plot_het(heterozygosity_maf_lower, params), ncol = 2) + 
  patchwork::plot_annotation(
    title = "Heterozygosity Rate",
    tag_levels = "A", 
    subtitle = paste0(
      "The horizontal red lines (\"<b style = 'color:firebrick2;'>---</b>\") denotes the ", params[["heterozygosity_threshold"]], 
      " times the standard deviations (SD) from the mean heterozygosity rate,<br>",
      "based on: <b>A</b>) all variants, <b>B</b>) variants with a MAF < ", params[["maf_threshold"]], "."
    )
  )
} else {
  patchwork::wrap_plots(plot_het(heterozygosity, params), ncol = 1) + 
  patchwork::plot_annotation(
    title = "Heterozygosity Rate", 
    subtitle = paste0(
      "The horizontal red lines (\"<b style = 'color:firebrick2;'>---</b>\") denotes the ", params[["heterozygosity_threshold"]], 
      " times the standard deviations (SD) from the mean heterozygosity rate, based on all variants."
    )
  )
}
```

```{r heterozygosity-tab, eval = nrow(heterozygosity[!is.na(labs)]) > 0}
heterozygosity[!is.na(labs), .(FID, IID, O.HOM., E.HOM., N.NM., hrate)] %>% 
  gt::gt(auto_align = "center") %>% 
  gt::tab_header(paste(
    "Samples To Be Excluded For An Heterozygosity Rate Higher Than", params[["heterozygosity_threshold"]], 
    "Times The Standard Deviations From The Mean Heterozygosity Rate."
  )) %>% 
  gt::opt_row_striping() %>% 
  gt::opt_all_caps() %>% 
  print()
```

## Relatedness

`r if (params[["includes_relatives"]]) cat("**Note:** Not applicable.\n\n")`

In order to identify pairs of samples who look too similar to each other, 
pairwise IBD (identity by descent) was computed, using the `--genome` command in PLINK.

Prior to the analysis, the dataset was pruned to exclude high-LD-regions, 
that is no variants with MAF < `r params[["maf_threshold"]]` and no pair of variants within 50 Kb have an r² > 0.2.

If the samples in the cohort **`r params[["cohort_name"]]`** are not supposed to be related, 
the expected pairwise $\hat{\pi}$ ("PI_HAT") should be less than `r params[["ibd_threshold"]]`.  
Pairs of sample with $\hat{\pi}$ ("PI_HAT") greater than `r params[["ibd_threshold"]]` were flagged.

```{r relatedness, eval = !params[["includes_relatives"]], include = FALSE}
## Pruning Data
sys(paste(plink, 
  "--bfile", file.path(temp, "data"), 
  "--autosome", 
  "--set-hh-missing", 
  "--maf", params[["maf_threshold"]], 
  "--r2", 
  "--ld-window-r2 0.2", 
  "--ld-window-kb 50", 
  "--make-bed",
  "--out", file.path(temp, "data_pruned")
))

sys(paste(plink, 
  "--bfile", file.path(temp, "data_pruned"), 
  "--freq", 
  "--out ", file.path(temp, "data_pruned")
))

sys(paste(plink, 
  "--bfile", file.path(temp, "data_pruned"), 
  "--read-freq", file.path(temp, "data_pruned.frq"), 
  "--genome", 
  "--min", params[["ibd_threshold"]],  
  "--out", file.path(temp, "data_pruned_freq")
))

relatedness <- unique(
  data.table::fread(
    file = file.path(temp, "data_pruned_freq.genome"), 
    colClasses = c(
      "FID1" = "character", "IID1" = "character", 
      "FID2" = "character", "IID2" = "character",
      "PI_HAT" = "numeric"
    )
  )
)
relatedness <- relatedness[!paste(FID1, IID1, sep = "_") %in% samples_to_exclude[, paste(FID, IID, sep = "_")], ]
relatedness <- relatedness[!paste(FID2, IID2, sep = "_") %in% samples_to_exclude[, paste(FID, IID, sep = "_")], ]
```

```{r relatedness-tab, eval = !params[["includes_relatives"]]}
relatedness_tab <- relatedness[, .N, by = cut(
  x = PI_HAT,
  breaks = c(params[["ibd_threshold"]], 0.3, 0.4, 0.6, 0.8, 1),
  include.lowest = TRUE
)][order(cut)]

relatedness_tab[, 
  cut := factor(cut, levels = levels(cut), labels = paste0(
    levels(cut),
    c(" = Second Degree Relatives", "", " = First Degree Relatives", "", " = MZ Twins/Duplicates")
  ))
]

if (nrow(relatedness_tab) > 0) {
  relatedness_tab %>% 
    gt::gt(auto_align = "center") %>% 
    gt::cols_label(cut = "IBD Windows", N = "Pair Count") %>% 
    gt::tab_header(paste0(
      'The Distribution Of "PI_HAT" (Proportion Of IBD) For Related Samples In The Dataset ',
      "(IBD ≥ ", params[["ibd_threshold"]], ")"
    )) %>% 
    gt::opt_row_striping() %>% 
    gt::opt_all_caps() %>% 
    print()
}

related_samples <- merge(
  x = relatedness[, .(FID1, IID1, FID2, IID2, PI_HAT)],
  y = callrate_samples[, .(FID, IID, F_MISS)],
  by.x = c("FID1", "IID1"),
  by.y = c("FID", "IID")
)
related_samples <- merge(
  x = related_samples,
  y = callrate_samples[, .(FID, IID, F_MISS)],
  by.x = c("FID2", "IID2"),
  by.y = c("FID", "IID"), 
  suffixes = c("1", "2")
)
related_samples[, Call_Rate1 := percent2(1 - F_MISS1)]
related_samples[, Call_Rate2 := percent2(1 - F_MISS2)]
related_samples <- related_samples[, .SD, .SDcols = !c("F_MISS1", "F_MISS2")][order(-PI_HAT), ]
related_samples[,
  IBD := cut(
    x = PI_HAT,
    breaks = c(params[["ibd_threshold"]], 0.3, 0.4, 0.6, 0.8, 1),
    include.lowest = TRUE
  )
]
related_samples[, 
  IBD := factor(IBD, levels = levels(IBD), labels = paste0(
    levels(IBD),
    c(" = Second Degree Relatives", "", " = First Degree Relatives", "", " = MZ Twins/Duplicates")
  ))
]

if (nrow(related_samples) != 0) {
  related_samples %>% 
    gt::gt(groupname_col = "IBD", auto_align = "center") %>% 
    gt::tab_header(paste(
      "Pairs Of Samples With A Proportion Of IBD Higher Than", params[["ibd_threshold"]]
    )) %>% 
    gt::opt_row_striping() %>% 
    gt::opt_all_caps() %>% 
    print()
} else {
  cat("\n\nNo pairs of samples found with a proportion of IBD higher than", params[["ibd_threshold"]], ".\n\n")
}

samples_to_exclude <- rbind(
  samples_to_exclude,
  unique(rbind(
    relatedness[PI_HAT >= params[["ibd_threshold"]], .(FID = FID1, IID = IID1)],
    relatedness[PI_HAT >= params[["ibd_threshold"]], .(FID = FID2, IID = IID2)]
  ))[, QC := "Relatedness"]
)
```

## Mendel Errors

The sample-wise Mendel error was estimated using the `--mendel` command in PLINK.  
The per-trio error rate threshold is `r params[["mendelian_samples"]]`.

```{r imendel-error}
if (!params[["includes_relatives"]]) cat("**Note:** Not applicable.\n\n")

imendel_file_check <- params[["includes_relatives"]]
```

```{r imendel-error-tab, eval = params[["includes_relatives"]]}
sys(paste(plink, 
  "--bfile", file.path(temp, "data"), 
  "--mendel",
  "--out ", file.path(temp, "data") 
))

sys(paste(plink, 
  "--bfile", file.path(temp, "data"), 
  "--extract", file.path(temp, "data_maf_geq.txt"), 
  "--mendel", 
  "--out", file.path(temp, "data_maf_geq")
))

if (is_maf_lower) {
  sys(paste(plink, 
    "--bfile", file.path(temp, "data"), 
    "--extract", file.path(temp, "data_maf_lower.txt"), 
    "--mendel", 
    "--out", file.path(temp, "data_maf_lower")
  ))
}

imendel_file_check <- file.exists(file.path(temp, "data.imendel"))

if (!imendel_file_check) {
  cat("Mendel error rates check was not performed due to missing informations.\n\n")
} else {
  mendel_errors <- data.table::fread(
    file = file.path(temp, "data.imendel"),
    colClasses = c("FID" = "character", "IID" = "character")
  ) 
  mendel_errors[, percentage := N / snps_number]
  mendel_errors[, labs := data.table::fifelse(percentage > params[["mendelian_samples"]], IID, NA_character_)]
  mendel_errors[, labs_fam := data.table::fifelse(percentage > params[["mendelian_samples"]], FID, NA_character_)]
  
  mendel_errors_geq <- data.table::fread(
    file = file.path(temp, "data_maf_geq.imendel"),
    colClasses = c("FID" = "character", "IID" = "character")
  )[, percentage := N / snps_number]
  prop_maf_geq <- sum(mendel_errors_geq[["percentage"]] > params[["mendelian_samples"]]) / 
    nrow(mendel_errors_geq)
  prop_all <- sum(mendel_errors[["percentage"]] > params[["mendelian_samples"]]) / 
    nrow(mendel_errors)
  prop_diff_maf_geq <- abs((prop_maf_geq - prop_all) / prop_all) >= 0.05
  
  if (is_maf_lower) {
    mendel_errors_lower <- data.table::fread(
      file = file.path(temp, "data_maf_lower.imendel"),
     colClasses = c("FID" = "character", "IID" = "character")
    )[, percentage := N / snps_number]
    prop_maf_lower <- sum(mendel_errors_lower[["percentage"]] > params[["mendelian_samples"]]) /
      nrow(mendel_errors_lower)
    prop_diff_maf_lower <- abs((prop_maf_lower - prop_all) / prop_all) >= 0.05
  }
  
  cat(
    "A total of ", comma(nrow(mendel_errors)), " samples ",
    "(from ", unique(mendel_errors[, FID]), " families)",
    " had a Mendel error rate higher than ", percent2(params[["mendelian_samples"]]), 
    " and were flagged.",
    "\n\n",
    sep = ""
  )
  
  prop_diff_maf <- (!is.na(prop_diff_maf_lower) & prop_diff_maf_lower) | 
    (!is.na(prop_diff_maf_geq) & prop_diff_maf_geq)
  if (prop_diff_maf) {
    cat("**Note:**\n\n")
    if (!is.na(prop_diff_maf_lower) & prop_diff_maf_lower) {
      cat(
        "    * the sample-wise Mendel error rate ratio was different for variants ",
        "with MAF < ", params[["maf_threshold"]], 
        " (", percent2(prop_maf_lower), ") ",
        "compared to the Mendel error rate ratio for all variants ",
        "(", percent2(prop_all), ").\n",
        sep = ""
      )
    }
    if (!is.na(prop_diff_maf_geq) & prop_diff_maf_geq) {
      cat(
        "    * the sample-wise Mendel error rate ratio was different for variants ",
        "with MAF > ", params[["maf_threshold"]], 
        " (", percent2(prop_maf_geq),") ",
        "compared to Mendel error rate ratio for all variants ",
        "(", percent2(prop_all), ").\n",
        sep = ""
      )
    }
    cat("\n")
  }
  
  cols_table <- c(
    "Mendel Error Rate", 
    "Samples to Exclude\n(All Variants)", 
    paste0("Samples to Exclude\n(Variants with MAF ≥ ", params[["maf_threshold"]], ")")
  )
  
  mendel_thresholds <- sort(unique(c(params[["mendelian_samples"]], 0.05, 0.03, 0.02, 0.01)), decreasing = TRUE)
  
  cbind.data.frame(
    percent2(mendel_thresholds),
    rowSums(sapply(mendel_errors[["percentage"]], ">", 1 - mendel_thresholds)),
    rowSums(sapply(mendel_errors_geq[["percentage"]], ">", 1 - mendel_thresholds))
  ) %>% 
    `colnames<-`(cols_table) %>% 
    gt::gt(auto_align = "center") %>% 
    gt::tab_header("Number Of Samples To Exclude Based On Mendel Error Rate Thresholds") %>% 
    gt::tab_style(
      style = gt::cell_fill(color = "dodgerblue", alpha = 0.5), 
      locations = gt::cells_body(
        columns = gt::vars(!!!cols_table), 
        rows = `Call Rate Threshold` == percent2(params[["callrate_samples"]])
      )
    ) %>% 
    gt::opt_row_striping() %>% 
    gt::opt_all_caps() %>% 
    print()
}
```

```{r imendel-error-fig, eval = params[["includes_relatives"]] & imendel_file_check}
ggplot2::ggplot(
  data = mendel_errors[order(percentage)],
  mapping = ggplot2::aes(
    x = seq_len(nrow(mendel_errors)), 
    y = percentage
  )
) +
  ggplot2::geom_point(
    colour = scales::viridis_pal(begin = 0.5, end = 0.5)(1), 
    shape = 1, 
    na.rm = TRUE
  ) +
  ggplot2::geom_hline(
    mapping = ggplot2::aes(yintercept = params[["mendelian_samples"]]), 
    colour = "firebrick2", 
    linetype = 2
  ) +
  ggrepel::geom_label_repel(
    data = ~ .x[!is.na(labs)],
    mapping = ggplot2::aes(label = labs),
    min.segment.length = ggplot2::unit(0, "lines"),
    force = 10, 
    fill = "white",
    colour  = "firebrick2", 
    segment.colour = "firebrick2", 
    size = 3, 
    na.rm = TRUE
  ) +
  ggplot2::scale_x_continuous(labels = comma, trans = "log10") +
  ggplot2::scale_y_continuous(
    breaks = function(x) unique(c(scales::breaks_extended()(x), params[["mendelian_samples"]])),
    labels = function(x) ifelse(x == params[["mendelian_samples"]], paste0("<b style='color:firebrick2;'>", percent1(x), "</b>"), percent1(x))
  ) +
  ggplot2::labs(
    x = "Number of Samples", 
    y = "Mendel Error Rate",
    title = "Sample Mendel Error Rate"
  )
```

```{r imendel-error-exclude, eval = params[["includes_relatives"]] & imendel_file_check}
mendel_errors_samples <- mendel_errors[percentage > params[["mendelian_samples"]], .(FID, IID)][, QC := "Sample_Mendel_Error"]

if (nrow(mendel_errors_samples) > 0) {
  mendel_errors_samples[, .(FID, IID, Sample_Mendel_Error)] %>% 
    gt::gt(auto_align = "center") %>% 
    gt::fmt_percent("Sample_Mendel_Error", incl_space = TRUE) %>% 
    gt::cols_label(Sample_Mendel_Error = "Sample Mendel Error Rate") %>% 
    gt::tab_header(paste(
      "Samples To Be Excluded For A Sample Mendel Error Rate Threshold Of", percent2(params[["callrate_samples"]]), 
    )) %>% 
    gt::opt_row_striping() %>% 
    gt::opt_all_caps() %>% 
    print()
}

samples_to_exclude <- rbind(
  samples_to_exclude, 
  mendel_errors_samples
)
```

## Principal Component Analysis

### Ancestry Inference

The data from the **`r params[["cohort_name"]]`** cohort was pruned using the following cut-offs:

* Variants call rate > 90.00 %.
* MAF ≥ `r params[["maf_threshold"]]`.
* HWE > `r scientific(params[["hwe_pvalue"]])`.
* Samples call rate > `r percent2(params[["callrate_samples"]])`.
* Heterozygosity rate within `r params[["heterozygosity_threshold"]]` times the standard deviation from the mean heterozygosity rate.

```{r ancestry-inference, include = FALSE}
data.table::fwrite(
  x = data.table::fread(file.path(temp, "data.bim"))[V1 %in% 1:22, .(V2)],
  file = file.path(temp, "data.txt"),
  quote = FALSE, 
  col.names = FALSE, 
  row.names = FALSE
)

samples_to_exclude_pca <- unique(samples_to_exclude[grepl("Sample_Call_Rate|Heterozygosity_Check$", QC), .(FID, IID)])

data.table::fwrite(
  x = samples_to_exclude_pca,
  file = file.path(temp, "samples_to_exclude_pca.txt"),
  sep = "\t", 
  row.names = FALSE, 
  col.names = FALSE, 
  quote = FALSE
)

if (nrow(samples_to_exclude_pca) > 0) {
  sys(paste(plink, 
    "--bfile", file.path(temp, "data"),
    "--remove", file.path(temp, "samples_to_exclude_pca.txt"),
    "--extract", file.path(temp, "data.txt"),
    "--snps-only",
    "--maf", params[["maf_threshold"]], 
    "--hwe", params[["hwe_pvalue"]], 
    "--geno 0.1", 
    "--make-bed",
    "--out", file.path(temp, "data_pop")
  ))
} else {
  sys(paste(plink, 
    "--bfile", file.path(temp, "data"), 
    "--extract", file.path(temp, "data.txt"),
    "--snps-only",
    "--maf", params[["maf_threshold"]], 
    "--hwe", params[["hwe_pvalue"]], 
    "--geno 0.1", 
    "--make-bed",
    "--out", file.path(temp, "data_pop")
  ))
}

data.table::fwrite(
  x = data.table::fread(file.path(temp, "data_pop.bim"))[
    !paste(V1, V4) %in% unique(paste(V1, V4)[duplicated(paste(V1, V4))]) &
      !V2 %in% unique(V2[duplicated(V2)]), 
    .(V2)
  ],
  file = file.path(temp, "clean_data_snp.txt"),
  row.names = FALSE, 
  col.names = FALSE, 
  quote = FALSE
)

invisible(parallel::mclapply(1:22, mc.cores = min(parallel::detectCores(), 22), function(i) {
  sys(paste(plink, 
    "--bfile", paste0(params[["ref1kg_genotypes"]], "Chr", i),
    "--extract", file.path(temp, "clean_data_snp.txt"),
    "--snps-only",
    "--make-bed",
    "--out", file.path(temp, file.path("1kgenome_pruned", sprintf("1kgenome_pruned_chr%02d_tmp", i)))
  ))
  data.table::fwrite(
    x = data.table::fread(file.path(temp, file.path("1kgenome_pruned", sprintf("1kgenome_pruned_chr%02d_tmp.bim", i))))[duplicated(V2), .(V2)],
    file = file.path(temp, file.path("1kgenome_pruned", sprintf("1kgenome_pruned_chr%02d_tmp.dupvar", i))),
    col.names = FALSE, 
    row.names = FALSE, 
    quote = FALSE
  )
  sys(paste(plink, 
    "--bfile", file.path(temp, file.path("1kgenome_pruned", sprintf("1kgenome_pruned_chr%02d_tmp", i))),
    "--exclude", file.path(temp, file.path("1kgenome_pruned", sprintf("1kgenome_pruned_chr%02d_tmp.dupvar", i))), 
    "--snps-only",
    "--make-bed", 
    "--out", file.path(temp, file.path("1kgenome_pruned", sprintf("1kgenome_pruned_chr%02d", i)))
  ))
}))

list_1kgenome <- file.path(temp, "1kgenome_pruned", "1kgenome_pruned_files.txt")
data.table::fwrite(
  x = data.table::data.table(file.path(temp, "1kgenome_pruned", sprintf("1kgenome_pruned_chr%02d", 1:22))),
  file = list_1kgenome,
  col.names = FALSE, 
  row.names = FALSE, 
  quote = FALSE
)
out <- file.path(temp, "1kgenome_pruned", "1kgenome_pruned_complete")
sys(paste(plink, "--merge-list", list_1kgenome, "--out", out))

if (file.exists(paste0(out, "-merge.missnp"))) {
  parallel::mclapply(1:22, mc.cores = min(parallel::detectCores(), 22), function(i) {
    sys(paste(plink, 
      "--bfile", paste0(params[["ref1kg_genotypes"]], "Chr", i),
      "--exclude", paste0(out, "-merge.missnp"),
      "--extract", file.path(temp, "clean_data_snp.txt"), 
      "--snps-only",
      "--make-bed",
      "--out", file.path(temp, "1kgenome_pruned", paste0("1kgenome_pruned_chr", sprintf("%02d", i)))
    ))
  })
  sys(paste(plink, 
    "--bfile", first_chromosome, 
    "--exclude", paste0(out, "-merge.missnp"), 
    "--merge-list", list_1kgenome, 
    "--snps-only",
    "--make-bed", 
    "--out", out
  ))
}

sys(paste(plink, 
  "--bfile", file.path(temp, "1kgenome_pruned", "1kgenome_pruned_complete"),
  "--bmerge", 
    file.path(temp, "data_pop.bed"), 
    file.path(temp, "data_pop.bim"),
    file.path(temp, "data_pop.fam"), 
  "--make-bed", 
  "--out", file.path(temp, "data_1kgenome")
))

sys(paste(plink, 
  "--bfile", file.path(temp, "1kgenome_pruned", "1kgenome_pruned_complete"),
  "--flip", file.path(temp, "data_1kgenome-merge.missnp"), 
  "--make-bed",
  "--out", file.path(temp, "1kgenome_pruned", "1kgenome_pruned_flipped")
))

sys(paste(plink, 
  "--bfile", file.path(temp, "1kgenome_pruned", "1kgenome_pruned_flipped"),
  "--bmerge", 
    file.path(temp, "data_pop.bed"), 
    file.path(temp, "data_pop.bim"),
    file.path(temp, "data_pop.fam"), 
  "--make-bed", 
  "--out", file.path(temp, "data_1kgenome")
))

sys(paste(plink, 
  "--bfile", file.path(temp, "1kgenome_pruned", "1kgenome_pruned_flipped"),
  "--exclude", file.path(temp, "data_1kgenome-merge.missnp"), 
  "--make-bed",
  "--out", file.path(temp, "1kgenome_pruned", "1kgenome_pruned_flipped_cleaned")
))

sys(paste(plink, 
  "--bfile", file.path(temp, "data_pop"),
  "--exclude", file.path(temp, "data_1kgenome-merge.missnp"), 
  "--make-bed",
  "--out", file.path(temp, "data_pop_cleaned")
))

sys(paste(plink, 
  "--bfile", file.path(temp, "1kgenome_pruned", "1kgenome_pruned_flipped_cleaned"),
  "--bmerge ", 
    file.path(temp, "data_pop_cleaned.bed"), 
    file.path(temp, "data_pop_cleaned.bim"),
    file.path(temp, "data_pop_cleaned.fam"), 
  "--make-bed", 
  "--out", file.path(temp, "data_1kgenome")
))

data.table::fwrite(
  x = data.table::data.table(intersect(
   data.table::fread(file.path(temp, "1kgenome_pruned", "1kgenome_pruned_flipped_cleaned.bim"))[["V2"]],
   data.table::fread(file.path(temp, "data_pop_cleaned.bim"))[["V2"]]
  )),
  file = file.path(temp, "data_1kgenome_snps.txt"),
  col.names = FALSE, 
  row.names = FALSE, 
  quote = FALSE
)
sys(paste(plink, 
  "--bfile", file.path(temp, "data_1kgenome"), 
  "--extract", file.path(temp, "data_1kgenome_snps.txt"), 
  "--make-bed ",
  "--out", file.path(temp, "data_1kgenome")
))

data.table::fwrite(
  x = data.table::fread(file.path(temp, "data_1kgenome.bim"))[paste0(V5, V6) %in% c("GC", "CG", "AT", "TA"), .(V2)],
  file = file.path(temp, "data_1kgenome_palindromic.txt"),
  row.names = FALSE, 
  col.names = FALSE, 
  quote = FALSE
)
sys(paste(plink, 
  "--bfile", file.path(temp, "data_1kgenome"), 
  "--exclude", file.path(temp, "data_1kgenome_palindromic.txt"), 
  "--make-bed ",
  "--out", file.path(temp, "data_1kgenome")
))

sys(paste(params[["bin_path"]][["gcta"]], 
  "--bfile", file.path(temp, "data_1kgenome"),
  "--make-grm-bin", 
  "--pca", params[["pca_components"]], 
  "--thread-num", min(parallel::detectCores(), 22),
  "--out", file.path(temp, "data_1kgenome")
))
sys(paste(params[["bin_path"]][["gcta"]], 
  "--grm", file.path(temp, "data_1kgenome"), 
  "--pca", params[["pca_components"]], 
  "--thread-num", min(parallel::detectCores(), 22), 
  "--out", file.path(temp, "data_1kgenome")
))

n_good_snp <- nrow(data.table::fread(file.path(temp, "data_pop.bim"), select = 1))
n_good_snp_1kg <- nrow(data.table::fread(file.path(temp, "data_1kgenome.bim"), select = 1))
n_samples <- nrow(data.table::fread(file.path(temp, "data_1kgenome.fam"), select = 1))
n_samples_1kg <- nrow(data.table::fread(file.path(temp, "1kgenome_pruned", "1kgenome_pruned_flipped_cleaned.fam"), select = 1))

data_1kg_eigenvec <- data.table::fread(file.path(temp, "data_1kgenome.eigenvec"))
data.table::setnames(data_1kg_eigenvec, names(data_1kg_eigenvec), c("FID", "IID", paste0("PC", seq_len(params[["pca_components"]]))))

data_pca <- rbind(
  data.table::fread(
    file = file.path(temp, "data_pop.fam"), 
    col.names = c("FID", "IID"),
    select = c(1, 2),
    colClasses = c("V1" = "character", "V2" = "character")
  )[, .(FID, IID, POP = params[["cohort_name"]])],
  data.table::fread(
    file = params[["ref1kg_panel"]], 
    fill = TRUE, 
    header = TRUE, 
    col.names = c("IID", "POP"),
    select = c(1, 2),
    colClasses = c("sample" = "character")
  )[, .(FID =  IID, IID, POP)]
)[
  IID %in% data.table::fread(
    file = file.path(temp, "data_1kgenome.fam"), 
    select = 2, 
    colClasses = "character"
  )[["V2"]]
]

data_pca <- merge(x = data_pca, y = data_1kg_eigenvec, by = c("FID", "IID"), all.x = TRUE)[, 
  POP := factor(
    x = POP, 
    levels = c(
      params[["cohort_name"]], 
      sort(setdiff(unique(POP), params[["cohort_name"]]), decreasing = TRUE)
    )
  )
][, .(FID, IID, Population = POP, x = PC1, y = PC2)]

data_pca <- merge(
  x = data_pca,
  y = data_pca[, lapply(.SD, mean), by = Population, .SDcols = c("x", "y")],
  by = "Population", 
  all.x = TRUE, 
  suffixes = c("", ".centroid")
)[, 
  dist := sqrt((x - x.centroid)^2 + (y - y.centroid)^2)
][,
  close := dist > (stats::quantile(dist, 0.75) + 1.5 * stats::IQR(dist)), 
  by = Population
]

centre_pca <- data_pca[
  Population != params[["cohort_name"]], 
  .SD, 
  .SDcols = c("FID", "IID", "Population", grep("centroid", names(data_pca), value = TRUE))
][, 
  (c("FID", "IID", "Population")) := lapply(.SD, as.character), 
  .SDcols = c("FID", "IID", "Population")
]

check_ethnicity <- merge(
  x = data.table::fread(
    file = file.path(temp, "data_pop.fam"), 
    col.names = c("FID", "IID"), 
    select = c("V1" = "character", "V2" = "character")
  ),
  y = data_pca[Population == params[["cohort_name"]]],
  by = c("FID", "IID"),
  all.x = TRUE
)

check_ethnicity[,
  closest_pop := centre_pca[["Population"]][which.min(sqrt(
    (x - centre_pca[["x.centroid"]])^2 + (y - centre_pca[["y.centroid"]])^2
  ))],
  by = seq_len(nrow(check_ethnicity))
]

pop1kg_description <- data.table::fread(
  file = params[["ref1kg_population"]], 
  select = c(1, 4, 5),
  col.names = c("POP", "SUPER_POP", "SUPER_POP_NAME")
)

if (is.null(params[["population"]])) {
  not_target_population <- check_ethnicity[0, ]
} else {
  not_target_population <- check_ethnicity[!closest_pop %in% c(
    params[["population"]],
    pop1kg_description[SUPER_POP %in% params[["population"]], POP]
  )]
}

not_target_population <- not_target_population[, QC := "Ethnicity"]
```
    
The resulting dataset (*i.e.*, **`r comma(n_samples-n_samples_1kg)`** samples and **`r comma(n_good_snp)`** variants) was merged with the data from the publicly available *1,000 Genomes project* database, and thinned down to only include not palindromic variants that were present in both dataset. 

A principal component analysis (PCA) was performed on the merged dataset (*i.e.*, **`r comma(n_samples)`** samples and **`r comma(n_good_snp_1kg)`** variants, including **`r comma(n_samples_1kg)`** from the *1,000 Genomes project*).

```{r ancestry-inference-fig}
data_pca_plot <- merge(
  x = data_pca,
  y = merge(
    x = check_ethnicity[, .(FID, IID, closest_pop)], 
    y = fam_data[, .(FID, IID, cohort)], 
    all = TRUE, 
    by = c("FID", "IID")
  ),
  all = TRUE,
  by = c("FID", "IID")
)[, 
  Population := as.character(Population)
][, 
  cohort := data.table::fifelse(is.na(cohort), "1,000 Genomes", cohort)
][, 
  closest_pop := data.table::fifelse(is.na(closest_pop), Population, closest_pop)
][,
  Population := data.table::fifelse(Population %in% params[["cohort_name"]], cohort, Population)
]

data_pca_plot <- merge(
  x = data_pca_plot, 
  y = pop1kg_description, 
  all = TRUE,
  by.x = "Population", 
  by.y = "POP"
)[,
  SUPER_POP := data.table::fifelse(is.na(SUPER_POP), Population, SUPER_POP)
][,
  SUPER_POP_NAME := data.table::fifelse(is.na(SUPER_POP_NAME), Population, SUPER_POP_NAME)
][
  !is.na(cohort)
]

p_superpop <- ggplot2::ggplot(
  data = data_pca_plot[order(cohort)], 
  mapping = ggplot2::aes(
    x = x, 
    y = y, 
    colour = stats::relevel(factor(SUPER_POP_NAME), ref = params[["cohort_name"]]), 
    fill = stats::relevel(factor(SUPER_POP_NAME), ref = params[["cohort_name"]]),
    shape = SUPER_POP_NAME %in% params[["cohort_name"]], 
    size = as.numeric(SUPER_POP_NAME %in% params[["cohort_name"]])
  )
) +
  ggplot2::geom_hline(yintercept = 0, linetype = 2) +
  ggplot2::geom_vline(xintercept = 0, linetype = 2) +
  ggplot2::geom_point(na.rm = TRUE, key_glyph = ggplot2::draw_key_rect) +
  ggforce::geom_mark_ellipse(
    data = ~ .x[SUPER_POP_NAME == params[["cohort_name"]]],
    alpha = 0.1,
    size = 0.5,
    con.cap = 0,
    con.border = "all",
    expand = ggplot2::unit(2, "mm"), 
    show.legend = FALSE
  ) +
  ggplot2::scale_colour_viridis_d(
    labels = function(x) ifelse(x == params[["cohort_name"]], paste0("**", params[["cohort_name"]], "**"), x)
  ) +
  ggplot2::scale_fill_viridis_d(
    labels = function(x) ifelse(x == params[["cohort_name"]], paste0("**", params[["cohort_name"]], "**"), x)
  ) +
  ggplot2::scale_size(range = c(0.25, 1)) +
  ggplot2::labs(
    x = "PC01", 
    y = "PC02", 
    colour = NULL, # "Super Population", 
    fill = NULL # "Super Population"
  ) +
  ggplot2::guides(
    shape = "none", 
    size = "none", 
    fill = ggplot2::guide_legend(
      ncol = (length(data_pca_plot[Population == params[["cohort_name"]], unique(closest_pop)]) > params[["max_labels"]]) + 1
    ),
    colour = "none"
  )

p_pop <- ggplot2::ggplot(
  data = data_pca_plot[Population %in% c(params[["cohort_name"]], data_pca_plot[Population == params[["cohort_name"]], unique(closest_pop)])][order(cohort)],
  mapping = ggplot2::aes(
    x = x, 
    y = y, 
    colour = stats::relevel(factor(Population), ref = params[["cohort_name"]]),
    fill = stats::relevel(factor(Population), ref = params[["cohort_name"]]), 
    shape = Population %in% params[["cohort_name"]], 
    size = as.numeric(Population %in% params[["cohort_name"]])
  )
) +
  ggplot2::geom_hline(yintercept = 0, linetype = 2) +
  ggplot2::geom_vline(xintercept = 0, linetype = 2) +
  ggplot2::geom_point(na.rm = TRUE, key_glyph = ggplot2::draw_key_rect) +
  ggforce::geom_mark_ellipse(
    data = ~ .x[Population == params[["cohort_name"]]],
    alpha = 0.1,
    size = 0.5,
    con.cap = 0,
    con.border = "all",
    expand = ggplot2::unit(2, "mm"), 
    show.legend = FALSE
  ) +
  ggplot2::scale_colour_viridis_d(
    labels = function(x) ifelse(x == params[["cohort_name"]], paste0("**", params[["cohort_name"]], "**"), x)
  ) +
  ggplot2::scale_fill_viridis_d(
    labels = function(x) ifelse(x == params[["cohort_name"]], paste0("**", params[["cohort_name"]], "**"), x)
  ) +
  ggplot2::scale_size(range = c(0.25, 1)) +
  ggplot2::labs(
    x = "PC01", 
    y = "PC02", 
    colour = NULL, # "Population", 
    fill = NULL # "Population"
  )  +
  ggplot2::guides(
    shape = "none", 
    size = "none", 
    fill = ggplot2::guide_legend(
      ncol = (length(data_pca_plot[Population == params[["cohort_name"]], unique(closest_pop)]) > params[["max_labels"]]) + 1
    ),
    colour = "none"
  )

p_inertia <- ggplot2::ggplot(
  data = data.table::fread(file.path(temp, "data_1kgenome.eigenval"))[, y := V1 / sum(V1)][1:params[["pca_components"]]][, x := 1:.N],
  mapping = ggplot2::aes(x = x, y = y)
) +
  ggplot2::geom_col(fill = scales::viridis_pal(begin = 0.5, end = 0.5)(1)) +
  ggplot2::scale_x_continuous(
    breaks = seq_len(params[["pca_components"]]),
    labels = function(x) sprintf("PC%02d", x), 
    guide = ggplot2::guide_axis(angle = 45)
  ) +
  ggplot2::scale_y_continuous(
    labels = percent2, 
    expand = ggplot2::expansion(mult = c(0, 0.05))
  ) +
  ggplot2::labs(
    x = "Principal Components", 
    y = "Contribution"
  )

patchwork::wrap_plots(p_inertia, p_superpop, p_pop, design = "12\n13") +
  patchwork::plot_annotation(
    title = "Principal Component Analysis (*Ancestry Inference*)",
    subtitle = paste(
      "First factorial plane computed with a combined dataset<sup>&dagger;</sup> involving samples from the 1,000 Genomes Project.", 
      'Samples were coloured by **B**) "super population" and **C**) "population".',
      sep = "<br>"
    ),
    caption = sprintf(
      "<sup>&dagger;</sup>%s variants for %s samples (including %s from the 1,000 Genomes Project).", 
      comma(n_good_snp_1kg), comma(n_samples), comma(n_samples_1kg)
    ),
    tag_levels = "A"
  )
```

```{r ancestry-inference-exclude, eval = !is.null(params[["population"]]) & nrow(not_target_population) > 0}
samples_to_exclude <- rbind(
  samples_to_exclude, 
  not_target_population[, .(FID, IID, QC)]
)

pop_name <- pop1kg_description[SUPER_POP %in% params[["population"]], unique(SUPER_POP_NAME)]

cat(
  "In total, ", comma(nrow(not_target_population)), " sample",
  data.table::fifelse(nrow(not_target_population) > 1, "s were", " was"),
  " determined as non-", pop_name, " based on PCA.\n", 
  sep = ""
)
not_target_population[, .(FID, IID, closest_pop)] %>%
  gt::gt(auto_align = "center") %>% 
  gt::cols_label(closest_pop = "Closest Population") %>% 
  gt::tab_header(paste0(
    "Samples Determined As Non-", pop_name, " Based On PCA ", 
    "(N = ", comma(nrow(not_target_population)), ")"
  )) %>% 
  gt::opt_row_striping() %>% 
  gt::opt_all_caps() %>% 
  print()

cat("\n")
```

### Cohort Structure

To assess the homogeneity of the **`r params[["cohort_name"]]`** cohort, a PCA was performed using only samples from the cohort.  
Samples were labelled and flagged as being possible outliers when the distance from the **`r params[["cohort_name"]]`** cohort centroid was greater than `r params[["pca_threshold"]]` times the standard deviation of all distances to the cohort centroid.

```{r cohort-structure, include = FALSE}
data.table::fwrite(
  x = samples_to_exclude[grepl("Sample_Call_Rate|Heterozygosity_Check$", QC), unique(.(FID, IID))],
  file = file.path(temp, "samples_to_exclude_pca_within.txt"),
  sep = "\t", 
  row.names = FALSE, 
  col.names = FALSE, 
  quote = FALSE
)

sys(paste(plink, 
  "--bfile", file.path(temp, "data_pop"), 
  "--remove", file.path(temp, "samples_to_exclude_pca_within.txt"), 
  "--make-bed", 
  "--out", file.path(temp, "data_pop_within_tmp0")
))
sys(paste(params[["bin_path"]][["gcta"]], 
  "--bfile", file.path(temp, "data_pop_within_tmp0"), 
  "--make-grm-bin", 
  "--thread-num", min(parallel::detectCores(), 22),
  "--out", file.path(temp, "data_pop_within_tmp1")
))
sys(paste(params[["bin_path"]][["gcta"]], 
  "--grm", file.path(temp, "data_pop_within_tmp1"), 
  "--pca", params[["pca_components"]],
  "--thread-num", min(parallel::detectCores(), 22),
  "--out", file.path(temp, "data_pop_within")
))

eigenvec_pca <- data.table::fread(
  file = file.path(temp, "data_pop_within.eigenvec"), 
  colClasses = c("V1" = "character", "V2" = "character")
)
data.table::setnames(
  x = eigenvec_pca, 
  old = names(eigenvec_pca), 
  new = c("FID", "IID", paste0("PC", seq_len(params[["pca_components"]])))
)
eigenvec_pca[,
  is_outlier := abs(PC1) > mean(PC1) + params[["pca_threshold"]] * stats::sd(PC1) | 
    abs(PC2) > mean(PC2) + params[["pca_threshold"]] * stats::sd(PC2) | 
    abs(PC3) > mean(PC3) + params[["pca_threshold"]] * stats::sd(PC3)
]

fam_data_pca <- merge(
  x = eigenvec_pca[, POP := params[["cohort_name"]]],
  y = fam_data[, .(FID, IID, cohort)],
  all.x = TRUE,
  by = c("FID", "IID")
)
```

Samples labelled and flagged as possible outliers: **`r comma(eigenvec_pca[, sum(is_outlier)])`**.

```{r cohort-structure-fig}
p_within <- lapply(list(c("PC1", "PC2"), c("PC1", "PC3"), c("PC2", "PC3")), function(ipcs) {
  ggplot2::ggplot(
    data = fam_data_pca[, IID := if (sum(is_outlier) > params[["max_labels"]]) NA else IID], 
    mapping = ggplot2::aes(
      x = .data[[ipcs[1]]], 
      y = .data[[ipcs[2]]], 
      colour = is_outlier, 
      fill = is_outlier,
      shape = is_outlier, 
      size = as.numeric(is_outlier)
    )
  ) +
    ggplot2::geom_hline(yintercept = 0, linetype = 2) +
    ggplot2::geom_vline(xintercept = 0, linetype = 2) +
    ggplot2::geom_point(data = ~ .x[!(is_outlier)], na.rm = TRUE) +
    ggforce::geom_mark_ellipse(
      data = ~ .x[!(is_outlier)],
      alpha = 0.1,
      size = 0.5,
      con.cap = 0,
      con.border = "all",
      expand = ggplot2::unit(2, "mm"), 
      show.legend = FALSE
    ) +
    ggplot2::geom_point(data = ~ .x[(is_outlier)], colour = "firebrick2", na.rm = TRUE) +
    ggrepel::geom_label_repel(
      data = ~ .x[(is_outlier)],
      mapping = ggplot2::aes(label = IID),
      min.segment.length = ggplot2::unit(0, "lines"),
      force = 10, 
      fill = "white",
      colour  = "firebrick2", 
      segment.colour = "firebrick2", 
      size = 3, 
      na.rm = TRUE
    ) +
    ggplot2::scale_colour_viridis_d(begin = 0.5, end = 0.5) +
    ggplot2::scale_fill_viridis_d(begin = 0.5, end = 0.5) +
    ggplot2::scale_size(range = c(0.75, 3)) +
    ggplot2::scale_shape_manual(values = c(1, 3)) +
    ggplot2::labs(x = gsub("PC", "PC0", ipcs[1]), y = gsub("PC", "PC0", ipcs[2])) +
    ggplot2::theme(legend.position = "none")
})

p_within[[length(p_within) + 1]] <- ggplot2::ggplot(
  data = data.table::fread(
    file = file.path(temp, "data_pop_within.eigenval")
  )[, 
    y := V1 / sum(V1)
  ][
    1:params[["pca_components"]]
  ][, 
    x := 1:.N
  ],
  mapping = ggplot2::aes(x = x, y = y)
) +
  ggplot2::geom_col(fill = scales::viridis_pal(begin = 0.5, end = 0.5)(1)) +
  ggplot2::scale_x_continuous(
    breaks = seq_len(params[["pca_components"]]),
    labels = function(x) sprintf("PC%02d", x), 
    guide = ggplot2::guide_axis(angle = 45)
  ) +
  ggplot2::scale_y_continuous(
    labels = percent2, 
    expand = ggplot2::expansion(mult = c(0, 0.05))
  ) +
  ggplot2::labs(
    x = "Principal Components", 
    y = "Contribution"
  )

patchwork::wrap_plots(p_within, nrow = 2, ncol = 2) +
  patchwork::plot_annotation(
    title = "Principal Component Analysis (*Cohort Structure*)",
    subtitle = paste(
      sprintf(
        "The PCA was computed on a dataset of %s variants for %s samples.<br>", 
        comma(nrow(data.table::fread(file.path(temp, "data_pop_within_tmp0.bim")))),
        comma(nrow(data.table::fread(file.path(temp, "data_pop_within_tmp0.fam"))))
      ),
      "Samples at more than", params[["pca_threshold"]], 
      "times the standard deviation from the mean of the principal components 1 to 3 are coloured",
      "and denoted by \"<b style = 'color:firebrick2;'>+</b>\"."
    ), 
    tag_levels = "A"
  )
```

```{r cohort-structure-exclude}
cohort_outliers <- eigenvec_pca[(is_outlier), .(FID, IID)][, QC := "Cohortwithin"]

if (nrow(cohort_outliers) > 0) {
  cohort_outliers[, .(FID, IID)] %>% 
    gt::gt(auto_align = "center") %>% 
    gt::tab_header("Outliers Within The Cohort") %>% 
    gt::opt_row_striping() %>% 
    gt::opt_all_caps() %>% 
    print()
}

samples_to_exclude <- rbind(
  samples_to_exclude,
  cohort_outliers
)
```

```{r summary-exclude, include = FALSE}
samples_to_exclude <- samples_to_exclude[, (c("FID", "IID")) := unique(.(FID, IID)), by = QC]
samples_to_exclude[, 
  QC := factor(
    x = QC,
    levels = c(
      "Sample_Call_Rate", 
      "Sex_Discrepancy", 
      "Sex_Missing", 
      "Sex_F", 
      "Heterozygosity_Check", 
      paste0("Heterozygosity_Check_maf_", params[["maf_threshold"]]),
      "Relatedness", 
      "Sample_Mendel_Error", 
      "Ethnicity", 
      "Cohortwithin"
    )
  )
]
samples_to_exclude[, value := 1]
samples_to_exclude_export <- data.table::dcast(samples_to_exclude, FID + IID ~ QC, value.var = "value", fill = 0)
if (!"Sample_Call_Rate" %in% names(samples_to_exclude_export)) samples_to_exclude_export[, Sample_Call_Rate := 0]
if (!"Heterozygosity_Check" %in% names(samples_to_exclude_export)) samples_to_exclude_export[, Heterozygosity_Check := 0]
samples_to_exclude_export[, Status := data.table::fifelse((Sample_Call_Rate + Heterozygosity_Check) > 0, "Exclude", "Check")]
name_cols <- c(
  "FID", "IID", "Status",
  "Call Rate", "Sex Discrepancy", "Sex Missing", "Sex Other", 
  paste0("Heterozygosity Check (MAF ", params[["maf_threshold"]], ")"),
  "Heterozygosity Check", "Ethnicity", "Cohort PCA outliers",
  "Relatedness Pi_hat ≥ 0.2", "Mendelian Error"
)
names(name_cols) <- c(
    "FID", "IID", "Status",
    "Sample_Call_Rate", "Sex_Discrepancy", "Sex_Missing", "Sex_F",
    paste0("Heterozygosity_Check_maf_", params[["maf_threshold"]]),
    "Heterozygosity_Check", "Ethnicity", "Cohortwithin", 
    "Relatedness", "Sample_Mendel_Error"
  )
data.table::setnames(
  x = samples_to_exclude_export, 
  old = names(samples_to_exclude_export), 
  new = name_cols[names(samples_to_exclude_export)]
)

data.table::fwrite(
  x = samples_to_exclude_export[Status == "Exclude"],
  file = file.path(temp, "samples_excluded.txt"),
  sep = "\t", 
  col.names = TRUE, 
  quote = FALSE, 
  row.names = FALSE
)

sys(paste(plink, 
  "--bfile", file.path(temp, "data"), 
  "--remove", file.path(temp, "samples_excluded.txt"), 
  "--snps-only", 
  "--set-hh-missing", 
  "--make-bed", 
  "--out", file.path(temp, "data_good_samples")
))
```

# Variant-based QC

Variant-based QC was performed on **`r comma(nrow(data.table::fread(file.path(temp, "data_good_samples.fam"), select = 1)))`** samples (*i.e.*, after excluding samples identified in the Section [Sample-based QC](#sample-qc)).  

## Variant Call Rate

Proportion of missing genotypes (i.e., *call rate*) per variant was estimated using the `--missing` command in PLINK.

```{r variant-call-rate}
sys(paste(plink, 
  "--bfile", file.path(temp, "data_good_samples"), 
  "--missing",
  "--out ", file.path(temp, "data_good_samples")
))
callrate_snps_raw <- data.table::fread(file.path(temp, "data_good_samples.lmiss"))
snps_low_call_rate <- callrate_snps_raw[, sum((1 - F_MISS) < params[["callrate_snps"]])]
```

In total, **`r comma(snps_low_call_rate)`** variants had a call rate below `r percent2(params[["callrate_snps"]])` and were excluded.  

```{r variant-call-rate-fig}
snps_call_rate <- callrate_snps_raw[, F_MISS := data.table::fifelse(is.na(F_MISS), 1, F_MISS)][F_MISS >= 0.001][order(1 - F_MISS)]
p <- ggplot2::ggplot(
  data = snps_call_rate,
  mapping = ggplot2::aes(x = seq_along(F_MISS), y = 1 - F_MISS)
) +
  ggplot2::geom_point(
    colour = scales::viridis_pal(begin = 0.5, end = 0.5)(1), 
    shape = 1, 
    na.rm = TRUE
  ) +
  ggplot2::geom_hline(
    mapping = ggplot2::aes(yintercept = params[["callrate_snps"]]),
    colour = "firebrick2", 
    linetype = 2
  ) +
  ggplot2::geom_vline(
    mapping = ggplot2::aes(xintercept = snps_low_call_rate),
    colour = "firebrick2", 
    linetype = 2
  ) +
  ggplot2::scale_x_continuous(
    labels = comma,
    trans = "log10",
    guide = ggplot2::guide_axis(check.overlap = TRUE), 
    sec.axis = ggplot2::dup_axis(
      labels = function(x) paste0("<b style = 'color:firebrick2;'>", comma(x), "</b>"),
      breaks = snps_low_call_rate, 
      name = NULL
    )
  ) +
  ggplot2::scale_y_continuous(
    breaks = function(x) unique(c(scales::breaks_extended()(x), params[["callrate_snps"]])),
    labels = function(x) data.table::fifelse(x == params[["callrate_snps"]], paste0("<b style='color:firebrick2;'>", percent1(x), "</b>"), percent1(x))
  ) +
  ggplot2::labs(
    x = "Number of Variants", 
    y = "Call Rate", 
    title = "Variants Call Rate",
    caption = "Variants with a call rate ≥ 99.9 % are not shown."
  )

suppressWarnings({ print(p) })
```

## Variants With Multiple Copies

```{r duplicated-snps, include = FALSE}
callrate_snps <- merge(
  x = callrate_snps_raw[, .(CHR, SNP, F_MISS)],
  y = data.table::fread(
    file = file.path(temp, "data_good_samples.bim"),
    select = c(1, 2, 4),
    col.names = c("CHR", "SNP", "POS")
  ),
  all.x = TRUE,
  by = c("CHR", "SNP")
)[order(CHR, POS)]
callrate_snps[, CHRPOS := paste(CHR, POS, sep = ":")]
callrate_snps[, (c("CHR", "POS")) := lapply(.SD, as.character), .SDcols = c("CHR", "POS")]

duplicated_snps <- callrate_snps[CHRPOS %in% CHRPOS[duplicated(CHRPOS)]]
uniq_snps <- callrate_snps[!CHRPOS %in% CHRPOS[duplicated(CHRPOS)]]
```

In total, **`r comma(nrow(duplicated_snps))`** variants existed in multiple copies in the dataset (same chromosome and position) and were removed.  

## Mendel Errors

```{r lmendel-error}
if (!params[["includes_relatives"]]) cat("**Note:** Not Applicable.\n\n")

lmendel_file_check <- params[["includes_relatives"]]

if (params[["includes_relatives"]]) {
  sys(paste(plink, 
    "--bfile", file.path(temp, "data_good_samples"), 
    "--mendel",
    "--out ", file.path(temp, "data_good_samples")
  ))
  
  lmendel_file_check <- file.exists(file.path(temp, "data_good_samples.lmendel"))
  
  cat(
    "The variant-wise Mendel error was estimated using the `--mendel` command in PLINK.  \n",
    "The per-variant error rate threshold is `", params[["mendelian_snps"]], "`.\n\n", 
    sep = ""
  )
  
  if (!lmendel_file_check) {
    cat("Mendel error rates check was not performed due to missing informations.\n\n")
  } else {
    mendel_error_snps <- data.table::fread(file.path(temp, "data_good_samples.lmendel"))[, percentage := N / .N][order(percentage)]
  }
}
```

```{r lmendel-error-fig, eval = params[["includes_relatives"]] & lmendel_file_check}
ggplot2::ggplot(data = mendel_error_snps, mapping = ggplot2::aes(x = seq_len(percentage), y = percentage)) +
  ggplot2::geom_point(
    colour = scales::viridis_pal(begin = 0.5, end = 0.5)(1), 
    shape = 1, 
    na.rm = TRUE
  ) +
  ggplot2::geom_hline(yintercept = params[["mendelian_snps"]], colour = "firebrick2", linetype = 2) +
  ggplot2::scale_x_continuous(labels = comma, trans = "log10") +
  ggplot2::scale_y_continuous(labels = percent1) +
  ggplot2::labs(
    x = "Number of Variants", 
    y = "Mendel Error Rate", 
    title = "Mendel Error Rate"
  )
```

```{r uniq-snps-list, include = FALSE}
if (lmendel_file_check) {
  uniq_snps_vec <- setdiff(uniq_snps[["SNP"]], mendel_error_snps[percentage > params[["mendelian_snps"]], SNP])
} else {
  uniq_snps_vec <- uniq_snps[["SNP"]]
}
cat(uniq_snps_vec, sep = "\n", file = file.path(temp, "uniq_snps_list.txt"))
```

## Minor Allele Frequency Distribution

The MAF were computed using the `--freq` command in PLINK.

```{r maf, include = FALSE}
sys(paste(plink, 
  "--bfile", file.path(temp, "data_good_samples"), 
  "--extract", file.path(temp, "uniq_snps_list.txt"),
  "--freq",
  "--out", file.path(temp, "data_good_samples")
))
```

```{r maf-fig}
ggplot2::ggplot(
  data = data.table::fread(file.path(temp, "data_good_samples.frq"))[!is.na(MAF)][,
    maf_bin := cut(MAF, breaks = c(0, 0.01, 0.05, 0.1, 0.2, 0.3, 0.4, 0.5), include.lowest = TRUE)
  ],
  mapping = ggplot2::aes(x = maf_bin)
) +
  ggplot2::geom_bar(fill = scales::viridis_pal(begin = 0.5, end = 0.5)(1)) +
  ggplot2::scale_y_continuous(labels = comma, expand = ggplot2::expansion(mult = c(0, 0.1))) +
  ggplot2::labs(
    x = "MAF", 
    y = "Number of Variants",
    title = "Distribution of variants number per class of minor allele frequency (MAF)"
  )
```

## Hardy-Weinberg Equilibrium

Hardy-Weinberg equilibrium (HWE) p-value was computed using the `--hardy` command in PLINK.

```{r hwe, include = FALSE}
sys(paste(plink, 
  "--bfile", file.path(temp, "data_good_samples"),
  "--hardy", 
  "--out ", file.path(temp, "data_good_samples")
))

hwe_complete <- data.table::fread(file.path(temp, "data_good_samples.hwe"))[!is.na(P)]
hwe_significant <- hwe_complete[P < 1e-3][order(P)]
n_snps_hwe <- hwe_complete[, sum(P < params[["hwe_pvalue"]])]
```

In total, **`r comma(n_snps_hwe)`** variants had a HWE p-value below `r scientific(params[["hwe_pvalue"]])` and were excluded.  

```{r hwe-fig, warning = FALSE}
p <- ggplot2::ggplot(data = hwe_significant, mapping = ggplot2::aes(x = 1:length(P), y = P)) +
  ggplot2::geom_point(
    colour = scales::viridis_pal(begin = 0.5, end = 0.5)(1), 
    shape = 1, 
    na.rm = TRUE
  ) +
  ggplot2::geom_hline(yintercept = params[["hwe_pvalue"]], colour = "firebrick2", linetype = 2) +
  ggplot2::geom_vline(xintercept = n_snps_hwe, colour = "firebrick2", linetype = 2) +
  ggplot2::scale_x_continuous(
    labels = comma,
    trans = "log10",
    guide = ggplot2::guide_axis(check.overlap = TRUE), 
    sec.axis = ggplot2::dup_axis(
      labels = function(x) paste0("<b style = 'color:firebrick2;'>", comma(x), "</b>"),
      breaks = n_snps_hwe, 
      name = NULL
    )
  ) +
  ggplot2::scale_y_continuous(trans = pval_trans(alpha = params[["hwe_pvalue"]], md = TRUE)) +
  ggplot2::labs(
    x = "Number of Variants", 
    y = "Hardy-Weinberg Equilibrium P-value",
    title = "Distribution of Hardy-Weinberg Equilibrium (HWE) P-values",
    subtitle = paste(
      "The horizontal red line (\"<b style = 'color:firebrick2;'>---</b>\") denotes the",
      "cut-off for significance above which variants are excluded."
    ),
    caption = paste(
      "Variants with a HWE p-value ≥ 10<sup>-3</sup> are not shown."
    )
  )

suppressWarnings({ print(p) })
```

```{r file-management, include = FALSE}
imputation_filename <- file.path(temp, paste0(params[["cohort_name"]], "_forimputation"))

sys(paste(plink, 
  "--bfile", file.path(temp, "data"), 
  "--set-hh-missing", 
  "--remove", file.path(temp, "samples_excluded.txt"),
  "--extract", file.path(temp, "uniq_snps_list.txt"), 
  "--hwe", params[["hwe_pvalue"]],
  "--make-bed", 
  "--out", paste0(imputation_filename, "_tmp1")
))

if (file.exists(paste0(imputation_filename, "_tmp1.hh"))) {
  sys(paste(plink, 
    "--bfile", paste0(imputation_filename, "_tmp1"), 
    "--exclude", paste0(imputation_filename, "_tmp1.hh"), 
    "--make-bed", 
    "--out", imputation_filename
  ))
} else {
  sys(paste(plink, 
    "--bfile", paste0(imputation_filename, "_tmp1"), 
    "--make-bed", 
    "--out", imputation_filename
  ))
}

sys(paste(plink, 
  "--bfile", imputation_filename, 
  "--geno", 1 - params[["callrate_snps"]],
  "--make-bed", 
  "--out", file.path(params[["output_directory"]], "plink_qc", paste0(params[["cohort_name"]], "_QC"))
))

sys(paste(plink, 
  "--bfile", imputation_filename, 
  "--freq", 
  "--out", imputation_filename
))

if (params[["imputation_ref"]] == "HRC") {
  system(paste(
    "(cd", temp, "&&",
      "perl", params[["check_bim_script"]],
      "-b", paste0(imputation_filename, ".bim"),
      "-f", paste0(imputation_filename, ".frq"),
      "-r", params[["imputation_panel"]], 
      "-h",
      "-l", plink,
    ")"
  ))
}

if (params[["imputation_ref"]] == "1KG") {
  system(paste(
    "(cd", temp, "&&",
      "perl", params[["check_bim_script"]],
      "-b", paste0(imputation_filename, ".bim"),
      "-f", paste0(imputation_filename, ".frq"),
      "-r", params[["imputation_panel"]], 
      "-g", 
      "--pop ALL",
      "-l", plink,
    ")"
  ))
}

system(paste("cd", temp, "&& bash Run-plink.sh"))

chr_exists <- file.exists(sprintf(paste0(imputation_filename, "-updated-chr%d.bim"), 1:23))
data.table::fwrite(
  x = data.table::data.table(sprintf(paste0(imputation_filename, "-updated-chr%d"), c(1:23)[which(chr_exists)])),
  file = file.path(temp, paste0(params[["cohort_name"]], "_forimputation.txt")),
  col.names = FALSE, 
  row.names = FALSE, 
  quote = FALSE
)

sys(paste(plink,
  "--merge-list", file.path(temp, paste0(params[["cohort_name"]], "_forimputation.txt")),
  "--recode vcf",
  "--out", imputation_filename
))

system(paste(params[["bin_path"]][["bgzip"]], 
  "-c", paste0(imputation_filename, ".vcf"),
  ">", paste0(imputation_filename, ".vcf.gz")
))
system(paste(params[["bin_path"]][["bcftools"]], "index", paste0(imputation_filename, ".vcf.gz")))

data.table::fwrite(
  x = structure(
    list(
      V1 = 1:26,
      V2 = structure(c(
        1L, 12L, 16L, 17L, 18L, 19L, 20L, 21L, 22L, 2L, 3L, 4L, 5L, 6L, 
        7L, 8L, 9L, 10L, 11L, 13L, 14L, 15L, 24L, 25L, 24L, 23L
      ),
      .Label = c(
        "1", "10", "11", "12", "13", "14", "15", "16", "17", "18", "19", "2",
        "20", "21", "22", "3", "4", "5", "6", "7", "8", "9", "MT", "X", "Y"
      ),
      class = "factor"
      )
    ),
    .Names = c("V1", "V2"), class = "data.frame", row.names = c(NA, -26L)
  ),
  file = file.path(temp, "plink2ensembl.txt"),
  row.names = FALSE, 
  col.names = FALSE, 
  quote = FALSE, 
  sep = "\t"
)

if (is.null(params[["ref1kg_fasta"]])) {
  system(paste(params[["bin_path"]][["bcftools"]], 
    "annotate", 
    "--rename-chrs", file.path(temp, "plink2ensembl.txt"),
    paste0(imputation_filename, ".vcf.gz"),
    "-Oz", 
    ">", file.path(params[["output_directory"]], "vcf_qc", paste0(params[["cohort_name"]], "_QC.vcf.gz"))
  ))
  
  system(paste(
    params[["bin_path"]][["bcftools"]], 
    "index", 
    file.path(params[["output_directory"]], "vcf_qc", paste0(params[["cohort_name"]], "_QC.vcf.gz"))
  ))
} else {
  system(paste(params[["bin_path"]][["bcftools"]], 
    "annotate", 
    "--rename-chrs", file.path(temp, "plink2ensembl.txt"),
    paste0(imputation_filename, ".vcf.gz"),
    "-Oz", 
    ">", file.path(params[["output_directory"]], "vcf_qc", paste0(params[["cohort_name"]], "_QC_tmp.vcf.gz"))
  ))
  system(paste(
    params[["bin_path"]][["bcftools"]], 
    "index", 
    file.path(params[["output_directory"]], "vcf_qc", paste0(params[["cohort_name"]], "_QC_tmp.vcf.gz"))
  ))
  system(paste(params[["bin_path"]][["bcftools"]],
    "+fixref", 
    file.path(params[["output_directory"]], "vcf_qc", paste0(params[["cohort_name"]], "_QC_tmp.vcf.gz")), 
    "-Ob -o", file.path(params[["output_directory"]], "vcf_qc", paste0(params[["cohort_name"]], "_QC.vcf.gz")),
    "--", 
    "-d", 
    "-f", params[["ref1kg_fasta"]], 
    "-m flip"
  ))
  system(paste(params[["bin_path"]][["bcftools"]], 
    "norm", 
    "--check-ref e", 
    "-f ", params[["ref1kg_fasta"]], 
    file.path(params[["output_directory"]], "vcf_qc", paste0(params[["cohort_name"]], "_QC.vcf.gz")), 
    "-o /dev/null"
  ))
  system(paste(params[["bin_path"]][["bcftools"]], 
    "index", 
    file.path(params[["output_directory"]], "vcf_qc", paste0(params[["cohort_name"]], "_QC.vcf.gz"))
  ))
}

unlink(file.path(params[["output_directory"]], "vcf_qc", paste0(params[["cohort_name"]], "_QC_tmp.vcf.gz")))
unlink(file.path(params[["output_directory"]], "vcf_qc", paste0(params[["cohort_name"]], "_QC_tmp.vcf.gz.csi")))
```

# QC Summary

```{r qc-summary, include = FALSE}
initial_snps_number <- nrow(data.table::fread(paste0(file.path(temp, "data"), ".bim"), select = 1))

final_snps_number <- nrow(data.table::fread(file.path(params[["output_directory"]], "plink_qc", paste0(params[["cohort_name"]], "_QC.bim")), select = 1))

initial_samples_number <- nrow(data.table::fread(paste0(file.path(temp, "data"), ".fam"), select = 1))

final_samples_number <- nrow(data.table::fread(file.path(params[["output_directory"]], "plink_qc", paste0(params[["cohort_name"]], "_QC.fam")), select = 1))

final_snps_imputation <- nrow(data.table::fread(paste0(imputation_filename, ".bim"), select = 1))
final_samples_imputation <- nrow(data.table::fread(paste0(imputation_filename, ".fam"), select = 1))
```

* Samples identified in the Section [Sample-based QC](#sample-qc):
    + **`r comma(nrow(samples_to_exclude_export[Status == "Exclude", unique(.(FID, IID))]))`** samples excluded due to failures in the call rate and/or the heterozygosity checks.
    + **`r comma(nrow(samples_to_exclude_export[Status == "Check", unique(.(FID, IID))]))`** samples flagged.  
        Those samples need to be checked in order to determine whether or not they have to be excluded.
* Variants with a call rate below `r percent2(params[["callrate_snps"]])` were removed.
* Duplicated variants based on chromosome and position were removed.
* Variants with a Hardy-Weinberg Equilibrium p-value at `r scientific(params[["hwe_pvalue"]])` were removed.
`r if (params[["includes_relatives"]] & lmendel_file_check) cat("* Variants with a Mendelian error rate below", params[["mendelian_snps"]], "were removed.")`

The initial dataset included **`r comma(initial_samples_number)`** samples and **`r comma(initial_snps_number)`** variants.  
The final (PLINK format) dataset includes **`r comma(final_samples_number)`** samples and **`r comma(final_snps_number)`** variants.  
The dataset for imputation includes **`r comma(final_samples_imputation)`** samples and **`r comma(final_snps_imputation)`** variants.

```{r qc-summary-table}
rbind(
  merge(
    x = fam_data[, .(FID, IID, cohort, sex)],
    y = samples_to_exclude_export[, .(FID, IID, Status)],
    all.y = TRUE,
    by = c("FID", "IID")
  )[, .N, by = c("cohort", "Status")],
  fam_data[, .(Status = "Clear", N = .N - nrow(samples_to_exclude_export[, unique(.(FID, IID))])), by = "cohort"]
)[, Pct := N / sum(N)][, Status := factor(Status, levels = c("Clear", "Check", "Exclude"))][order(Status), .(Status, N, Pct)] %>% 
  gt::gt(rowname_col = "Status", auto_align = "center") %>% 
  gt::tab_header(
    title = "Quality-Control Summary",
    subtitle = gt::md(paste0("*",  params[["cohort_name"]], "*"))
  ) %>% 
  gt::cols_label(Pct = "Percentage") %>% 
  gt::fmt_percent(columns = "Pct", decimals = 2, incl_space = TRUE) %>% 
  gt::grand_summary_rows(columns = "Pct", fns = list(Total = ~ sum(.)), formatter = gt::fmt_percent, incl_space = TRUE) %>% 
  gt::grand_summary_rows(columns = "N", fns = list(Total = ~ sum(.)), formatter = gt::fmt_number, decimals = 0) %>% 
  gt::opt_row_striping() %>% 
  gt::opt_all_caps() %>% 
  print()
```

```{r output-exclusion, include = FALSE}
output_xlsx <- list()

exclude_flag <- merge(
  x = samples_to_exclude_export,
  y = fam_data[, .(FID, IID, Cohort = cohort, Sex = sex)],
  all.x = TRUE,
  by = c("FID", "IID")
)
data.table::setcolorder(
  x = exclude_flag, 
  neworder = c("FID", "IID", "Sex", "Cohort", "Status", setdiff(names(exclude_flag), c("FID", "IID", "Sex", "Cohort", "Status")))
)
cols_order <- setdiff(names(exclude_flag), c("FID", "IID", "Sex", "Cohort"))
setorderv(exclude_flag, cols_order, -1)
output_xlsx[["exclude_flag"]] <- exclude_flag

if (exists("related_samples") && nrow(related_samples) > 0) {
  output_xlsx[["relatedness"]] <- merge(
    x = merge(
      x = related_samples,
      y = fam_data[, .(FID, IID, Cohort = cohort)],
      all.x = TRUE,
      by.x = c("FID1", "IID1"),
      by.y = c("FID", "IID")
    ),
    y = fam_data[, .(FID, IID, Cohort = cohort)],
    all.x = TRUE,
    by.x = c("FID2", "IID2"),
    by.y = c("FID", "IID"),
    suffix = c("1", "2")
  )
}

if (exists("mendel_errors_samples") && nrow(mendel_errors_samples) > 0) {
  output_xlsx[["Sample_Mendel_Error_Rate"]] <- mendel_errors_samples[,
    ("Sample Mendel error rate") := percent2(percent)
  ][, -c("percent")]
}

if (exists("not_target_population") && nrow(not_target_population) > 0) {
  output_xlsx[["ethnicity"]] <- not_target_population[, .(FID, IID, closest_pop)]
}

writexl::write_xlsx(output_xlsx, path = file.path(params[["output_directory"]], "QC_exclusion.xlsx"))
unlink(temp, recursive = TRUE)
```

# Technical Information

## Softwares

```{r softwares-version, include = FALSE}
suppressWarnings({
  plink_ver <- strsplit(system(paste(plink, "--version"), intern = TRUE)[1], split = " ")[[1]][2]
  gcta_ver <- paste(strsplit(system(params[["bin_path"]][["gcta"]], intern = TRUE)[3], split = " ")[[1]][3:4], collapse = "-")
  bgzip_ver <- strsplit(system(paste(params[["bin_path"]][["bgzip"]], "--version"), intern = TRUE)[1], split = " ")[[1]][3]
  bcftools_ver <- strsplit(system(paste(params[["bin_path"]][["bcftools"]], "--version"), intern = TRUE)[1], split = " ")[[1]][2]
  check_bim_script_ver <- readLines(params[["check_bim_script"]], warn = FALSE) %>% 
    grep("^#", ., value = TRUE) %>% 
    grep("-v[0-9]", ., value = TRUE) %>% 
    gsub("# *-v", "", .) %>% 
    gsub(" ", "", .) %>% 
    package_version() %>% 
    sort(decreasing = TRUE) %>% 
    .[1]
})
```

* plink `` `r plink_ver` `` (Chang et al., 2015; doi:10.1186/s13742-015-0047-8)
* gcta `` `r gcta_ver` `` (Yang et al., 2011; doi:10.1016/j.ajhg.2010.11.011)
* bgzip `` `r bgzip_ver` `` ([http://www.htslib.org/](http://www.htslib.org/))
* bcftools `` `r bcftools_ver` `` ([http://www.htslib.org/](http://www.htslib.org/))
* `r basename(params[["check_bim_script"]])` `` `r check_bim_script_ver` `` (William Rayner, [https://www.well.ox.ac.uk/~wrayner/tools/](https://www.well.ox.ac.uk/~wrayner/tools/))

## R Session Information

```{r session-info, results = "markup"}
options("width" = 110)
sessioninfo::session_info()
```
